### PR TITLE
feat(fn): basic sending to Amplitude with no error handling

### PR DESCRIPTION
**Pull request by philbooth**
This is my first stab at a logging 2.0 function that does basic per-event sending to Amplitude. Tested against the `FxAccts_Dev` project using the fixture data in this PR, it works okay as far as it goes. But it doesn't try to do any error handling yet because I'm still not sure about a couple of things.

Firstly, regarding rate-limits, [the docs say](https://amplitude.zendesk.com/hc/en-us/articles/204771828-HTTP-API#upload-limit):

> Limit your upload to 100 batches/sec and 1000 events/sec. You can batch events into an upload but we recommend not sending more than 10 events per batch. Thus, we expect at most 100 batches to be sent per second, and so the 1000 events/sec limit still applies as we do not recommend sending more than 10 events per batch. Note: This applies to customers on our Starter Plan only. 
>
> For paying customers, please reach out to us if you need to send more than 1000 events/sec. There is no hard limit on the Amplitude Enterprise plan. However, please note that we will throttle requests for individual devices that exceed 60 events/second as stated in the below HTTP status codes section.

I take a couple of things from this:

1. Assuming that doing so is straightforward, it makes sense for us to batch events in to groups of 10. @jbuck, is it possible for logging 2.0 to invoke us in batches of 10 log lines rather than for every line individually?

2. Based on some very sketchy back-of-a-fag-packet figuring-out, I think we'll definitely hit the rate limit in prod if we don't batch events, but we probably won't if we send batches of 10. Either way, I guess it's a good idea to begin that conversation with them about rate limits on the Enterprise plan. @davismtl?

Next, for sane error handling, we really want to include an `insert_id` with each request, but doing that requires statefulness and I'm not sure how to approach that within the context of logging 2.0 / AWS Lambda. Also, if we start wading into murkier waters of retry logic and exponential backoff, I'm not sure what limits there are for the lifetime of a logging 2.0 function invocation and how we should handle any timeouts arriving thereof. Do we need to guarantee all data is sent to Amplitude or is some level of lossage acceptable? @jbuck, @rfk, any wisdom to impart here?

Finally, code reviews for what's here so far are welcome, @mozilla/fxa-devs r?

### CODE REVIEW

**Comment by philbooth** [comment]
(and apologies if creating a fresh repo was the wrong thing to do, I wasn't sure where the best place to put this was)

File: amplitude.py [hunk 1]
```diff
@@ -1,20 +1,35 @@
...
 
         assert("time" in event)
         assert("user_id" in event or "device_id" in event)
 
+        if "user_id" in event:
+            assert(event["user_id"] == event["user_properties"]["fxa_uid"])
+
+            user_id_hmac = hmac.new(HMAC_KEY, event["user_id"], hashlib.sha256)
+            event["user_id"] = user_id_hmac.hexdigest()
+            event["user_properties"]["fxa_uid"] = event["user_id"]
```
Comment (philbooth) [inline amplitude.py]: If we're doing this here, it's redundant to set `user_properties\.fxa_uid` in our logging code in the auth and content servers. We should remove that if/when this sticks.

Comment (philbooth) [reply amplitude.py]: (and get rid of the assertion on line 19 too, of course)

**Comment by rfk** [comment]
> we really want to include an insert_id with each request

Can we calculate this deterministically, e.g. as SHA256 of the event's JSON payload with keys in sorted order?

I think we should definitely include an `insert_id` so that we can more easily recover from failures, safely do backfill, etc.

**Comment by rfk** [comment]
> Also, if we start wading into murkier waters of retry logic and exponential backoff,
> I'm not sure what limits there are for the lifetime of a logging 2.0 function invocation
> and how we should handle any timeouts arriving thereof.

In my ideal world, we push this down into the infrastructure - if it invokes this function for some events in the log stream, and this function does not return successfully, then it trys again later than than swallowing the error.

But I've no idea how that maps on to logging2.0's capabilities in practice :-)

> Do we need to guarantee all data is sent to Amplitude or is some level of lossage acceptable?

I think we should aim for as close to such a guarantee as possible, including giving ourselves the option to re-publish chunks of time as a background process if we think there was an operational issue.

File: amplitude.py (hunk 1 above)
Comment (philbooth) [reply amplitude.py]: > ...it's redundant to set `user_properties.fxa_uid`...

@davismtl raised this independently, we can get rid of it entirely.

**Comment by philbooth** [comment]
> In my ideal world, we push this down into the infrastructure - if it invokes this function for some events in the log stream, and this function does not return successfully, then it trys again later than than swallowing the error.

Agreed, this would be a nice and simple approach for us. @jbuck, do you know if we can operate like this? And if so, any idea, what the correct return/raise semantics are for signalling errors to the calling code?

**Comment by jbuck** [comment]
> Assuming that doing so is straightforward, it makes sense for us to batch events in to groups of 10. @jbuck, is it possible for logging 2.0 to invoke us in batches of 10 log lines rather than for every line individually?

You'll have full control over this - @relud's preferred method of delivery will be sending an S3 event notification when a file is stored on S3 that [looks like this](http://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html). The script will download the file from S3, decompress it, and then send it to your transform function. You can copy the code from [the PII scrubber](https://github.com/mozilla-services/cloudops-deployment/blob/master/projects/fxa/logging/ansible/templates/transform.yml#L70-L92) but I think doing the s3-specific handling in a separate file might make the most sense to make local dev/testing more easy.

> I think we should definitely include an insert_id so that we can more easily recover from failures, safely do backfill, etc.

I think we could use the ETag from the S3 event notification - that's the hash of the file on S3 which should be unique enough right?

> Agreed, this would be a nice and simple approach for us. @jbuck, do you know if we can operate like this? And if so, any idea what the correct return/raise semantics are for signalling errors to the calling code?

Yes I believe we can do this. You can send S3 event notifications to SQS and let it call the lambda function when a new one arrives. This will give us built-in retry mechanisms. In terms of signaling that something failed, it looks like [raising an exception](http://docs.aws.amazon.com/lambda/latest/dg/python-exceptions.html) is the preferred method.

**Comment by philbooth** [comment]
Some more discussion from IRC:

```
17:58:50 <jbuck> pb: lmk if my comments on the amplitude PR make sense
17:59:04 <jbuck> if not I can clarify :)
17:59:16 <pb> jbuck: thanks, reading now
18:00:58 <jbuck> jrgm and I have logging 2.0 in a state where we're happy to leave it running, so next week we can work with relud to integrate the amplitude bits
18:01:12 <jbuck>  / do a backfill so we've got 2 weeks of data for the demo
18:02:31 <pb> jbuck: cool, sounds fabulous
18:03:56 <pb> jbuck: just on your first paragraph quickly, do i understand correctly that the argument to transform will be the entire contents of a file from s3?
18:04:18 <pb> jbuck: that's cool if so, just wondering what time period it will cover / how many lines to expect
18:05:03 <pb> jbuck: (and sorry if this is explained by the pii scrubber link, i don't have access to it)
18:05:16 <jbuck> pb: the lambda function will recieve an s3 event notification that looks like http://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html
18:05:26 <jbuck> the amplitude lambda fn that is
18:05:48 <jbuck> so, it's gotta do the "download file from s3, decompress, and then pass it to the transform function" fance
18:05:49 <pb> jbuck: oh wait, i get it. so i read from s3 myself?
18:05:55 <jbuck> yeahy
18:06:04 <jbuck> let me see what the typical size looks like
18:06:12 <jbuck> this would be the fxa-content server stream right?
18:06:25 <pb> jbuck: content and auth servers
18:06:50 <pb> jbuck: ideally filtered to lines where op=amplitudeEvent, but i can do that myself if not
18:07:06 <jbuck> hmm, you know, we might be able to filter
18:07:09 <jbuck> I'll check with relud
18:07:44 <pb> 👍
18:07:54 <jbuck> w/o filtering though, looks like 3.5MB/minute for auth and 0.5MB/minute for content
18:08:25 <pb> jbuck: so i'd get invoked every minute?
18:08:27 <jbuck> let me see how many amplitude events are in a given minute...
18:08:29 <jbuck> yeah
18:08:38 <pb> jbuck: awesome, this sounds perfect!
18:08:48 <jbuck> I think kinesis does roll-up by the minute or by 100MB
18:08:51 <jbuck> whatever comes first
18:10:15 <jbuck> 15383 events for the auth server
18:11:06 <jbuck> 1343 for the content server
18:13:47 <jbuck> pb: hm, does this actually mean we'd go over the limit for fxa-auth...
18:14:08 <pb> jbuck: yep. amplitude limits to 1000 events per second by default, so i definitely need to get ADavis to speak to them about upping the limit for our api key
18:14:35 <jbuck> or I guess we'd need to do some proactive throttling on our part to only send say 500/sec
18:14:54 <pb> jbuck: pfff, throttling is for chumps
18:14:59 <jbuck> lol
```

**Comment by philbooth** [comment]
Closing this, will open a fresh PR as the code has deviated significantly from the initial description and discussions now.

**Pull request closed without merging**